applications/asset_tracker_v2/src/modules/cloud_module.c

```diff
@@ -790,54 +804,156 @@ static void on_sub_state_cloud_connected(struct cloud_msg_data *msg)
 	}
 
 	if (IS_EVENT(msg, data, DATA_EVT_AGPS_REQUEST_DATA_SEND)) {
-		message_add(msg->module.data.data.buffer.buf,
-			    msg->module.data.data.buffer.len,
-			    AGPS_REQUEST,
-			    QOS_FLAG_RELIABILITY_ACK_REQUIRED,
-			    true);
+
+		if (IS_ENABLED(CONFIG_LWM2M_INTEGRATION)) {
+
+			char *paths[CONFIG_CLOUD_CODEC_LWM2M_PATH_LIST_ENTRIES_MAX];
+
+			__ASSERT(ARRAY_SIZE(paths) ==
+				 ARRAY_SIZE(msg->module.data.data.buffer.paths),
+				 "Path object list not the same size");
+
+			for (int i = 0; i < ARRAY_SIZE(paths); i++) {
+				paths[i] = msg->module.data.data.buffer.paths[i];
+			}
+
+			err = cloud_wrap_agps_request_send(
+						   NULL,
+						   msg->module.data.data.buffer.valid_object_paths,
+						   true,
+						   0,
+						   paths);
+			if (err) {
+				LOG_WRN("cloud_wrap_agps_request_send, err: %d", err);
+			}
+
+			return;
+		}
+
+		message_add_to_qos_library(msg->module.data.data.buffer.buf,
+					   msg->module.data.data.buffer.len,
+					   AGPS_REQUEST,
+					   QOS_FLAG_RELIABILITY_ACK_REQUIRED,
+					   true);
 	}
 
 	if (IS_EVENT(msg, data, DATA_EVT_DATA_SEND)) {
-		message_add(msg->module.data.data.buffer.buf,
-			    msg->module.data.data.buffer.len,
-			    GENERIC,
-			    QOS_FLAG_RELIABILITY_ACK_DISABLED,
-			    true);
+
+		if (IS_ENABLED(CONFIG_LWM2M_INTEGRATION)) {
+
+			char *paths[CONFIG_CLOUD_CODEC_LWM2M_PATH_LIST_ENTRIES_MAX];
+
+			__ASSERT(ARRAY_SIZE(paths) ==
+				 ARRAY_SIZE(msg->module.data.data.buffer.paths),
+				 "Path object list not the same size");
+
+			for (int i = 0; i < ARRAY_SIZE(paths); i++) {
+				paths[i] = msg->module.data.data.buffer.paths[i];
+			}
+
+			err = cloud_wrap_data_send(NULL,
+						   msg->module.data.data.buffer.valid_object_paths,
+						   true,
+						   0,
+						   paths);
+			if (err) {
+				LOG_WRN("cloud_wrap_data_send, err: %d", err);
```

Comment:
For the MQTT based integration layers `cloud_wrap_data_send` (and the other `_sends`) basically translates to a socket `send` call. If send fails there is usually an indication that the modem does not accept the data from the app due to not being able to empty its internal TLS buffer, which can be due to poor network conditions. This is likely to happen, and it does it shouldn't be treated as an irrecoverable error.

We could pimp this logic by having a set of aligned error codes that are sent from the various integration layers and then switch on those in the cloud module and log based on the severity of the received error code.

Reply:
Ill close "duplicates" of the same issues to tidy up a bit.

Reply:
This is inside a `if (IS_ENABLED(CONFIG_LWM2M_INTEGRATION))` and `cloud_wrap_data_send` translates to `lwm2m_engine_send`. Where most errors seem to be because the developer is doing something wrong.

Reply:
Right, ill make sure to change these to log_err

Reply:
@gregersrygg Fixed.